### PR TITLE
map: Avoid calling f on zero unless necessary

### PR DESCRIPTION
**Pull request by eschnett**
Closes https://github.com/JuliaSparse/SparseArrays.jl/issues/221.

### CODE REVIEW

File: src/higherorderfns.jl [hunk 1]
```diff
@@ -156,14 +156,18 @@ map!(f::Tf, C::AbstractSparseMatrixCSC, A::AbstractSparseMatrixCSC, Bs::Vararg{S
 map!(f::Tf, C::SparseVecOrMat, A::SparseVecOrMat, Bs::Vararg{SparseVecOrMat,N}) where {Tf,N} =
     (_checksameshape(C, A, Bs...); _noshapecheck_map!(f, C, A, Bs...))
 function _noshapecheck_map!(f::Tf, C::SparseVecOrMat, A::SparseVecOrMat, Bs::Vararg{SparseVecOrMat,N}) where {Tf,N}
-    fofzeros = f(_zeros_eltypes(A, Bs...)...)
-    fpreszeros = _iszero(fofzeros)
+    haszeros = _haszeros(C) && _haszeros(A) && all(_haszeros, Bs)
```
Comment (SobhanMP) [inline src/higherorderfns.jl:159]: why is there a `_haszero(C)`?.

File: src/higherorderfns.jl [hunk 2]
```diff
@@ -156,14 +156,18 @@ map!(f::Tf, C::AbstractSparseMatrixCSC, A::AbstractSparseMatrixCSC, Bs::Vararg{S
...
+    # Avoid calculating f(zero) unless necessary (it may fail)
+    fofzeros = haszeros ? f(_zeros_eltypes(A, Bs...)...) : nothing
+    fpreszeros = !haszeros || _iszero(fofzeros)
```
Comment (SobhanMP) [inline src/higherorderfns.jl:162]:
```suggestion
   !haszero || _iszero(f(_zeros_eltypes(A, Bs...)...))
```
what you wrote is type unstable.

File: src/higherorderfns.jl (hunk 1 above)
Comment (SobhanMP) [inline src/higherorderfns.jl:159]:
```suggestion
    haszeros = _haszeros(A) && all(_haszeros, Bs)
```

**Comment by SobhanMP** [comment]
this should make it type stable. feel free to revert my commit if you think you have a better solution/my concerns are baseless, wanted to make one of those click to commit thingy but could not figure out how to remove multiple lines.

**Comment by eschnett** [comment]
@SobhanMP Thank you. I thought the compiler would optimize away the type instability. If there's a chance it doesn't, then your code is the way to go.

**Comment by SobhanMP** [comment]
can it really do that?

**Comment by eschnett** [comment]
It works for iterators, where the `next` function either returns the next state or `nothing`. Given that iterators are ubiquitous, I assume that this optimizations is fairly robust. To my knowledge, this only works in cases where there is one non-trivial type and one (or more?) empty types (such as `Nothing`).

No need to be gallant here, let's go with your version.

**Comment by Wimmerer** [comment]
I believe that's a case of union-splitting right? I believe the limit for union-splitting is around 4.

**Comment by SobhanMP** [comment]
that's pretty neat.

**Comment by Wimmerer** [comment]
This will only solve the issue for "dense" sparse arrays right?

**Comment by SobhanMP** [comment]
it's possible that it still gets called unnecessarily if there there is no matching zeros (i.e. [1, 0] + [0, 1]). yeah.

**Review by Wimmerer — APPROVED** [review]
Relatively niche fix, since we still don't support types without a good `zero` for truly sparse arrays.

**Pull request merged**